File: server/libs/storage-redis/src/main/java/com/sharetimer/storage/redis/config/ReactiveRedisTemplateFactory.java
package com.sharetimer.storage.redis.config;

import org.springframework.data.redis.connection.RedisStandaloneConfiguration;
import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory;
import org.springframework.data.redis.core.ReactiveRedisTemplate;
import org.springframework.data.redis.serializer.RedisSerializationContext;
import org.springframework.data.redis.serializer.StringRedisSerializer;
import org.springframework.stereotype.Component;

import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class ReactiveRedisTemplateFactory {

  private final RedisProps redisProps;
  private volatile ReactiveRedisTemplate<String, String> template;

  public ReactiveRedisTemplate<String, String> getTemplate() {
    if (template == null) {
      synchronized (this) {
        if (template == null) {
          template = createTemplate();
        }
      }
    }
    return template;
  }

  private ReactiveRedisTemplate<String, String> createTemplate() {
    RedisStandaloneConfiguration redisConfig = new RedisStandaloneConfiguration(redisProps.getHost(),
        redisProps.getPort());

    LettuceConnectionFactory factory = new LettuceConnectionFactory(redisConfig);
    factory.afterPropertiesSet();

    RedisSerializationContext<String, String> serializationContext = RedisSerializationContext
        .<String, String>newSerializationContext(new StringRedisSerializer())
        .key(new StringRedisSerializer()).value(new StringRedisSerializer())
        .hashKey(new StringRedisSerializer()).hashValue(new StringRedisSerializer()).build();

    return new ReactiveRedisTemplate<>(factory, serializationContext);
  }
}

File: server/libs/storage-redis/src/main/kotlin/com/sharetimer/storage/redis/config/ReactiveRedisTemplateFactory.kt
package com.sharetimer.storage.redis.config

import org.springframework.data.redis.connection.RedisStandaloneConfiguration
import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory
import org.springframework.data.redis.core.ReactiveRedisTemplate
import org.springframework.data.redis.serializer.RedisSerializationContext
import org.springframework.data.redis.serializer.StringRedisSerializer
import org.springframework.stereotype.Component

@Component
class ReactiveRedisTemplateFactory(
    private val redisProps: RedisProps,
) {
    val template: ReactiveRedisTemplate<String, String> by lazy { createTemplate() }

    private fun createTemplate(): ReactiveRedisTemplate<String, String> {
        val redisConfig = RedisStandaloneConfiguration(redisProps.host, redisProps.port)

        val factory =
            LettuceConnectionFactory(redisConfig).apply {
                afterPropertiesSet()
            }

        val serializer = StringRedisSerializer()
        val serializationContext =
            RedisSerializationContext
                .newSerializationContext<String, String>(serializer)
                .key(serializer)
                .value(serializer)
                .hashKey(serializer)
                .hashValue(serializer)
                .build()

        return ReactiveRedisTemplate(factory, serializationContext)
    }
}
